crates/rwasm/machine/src/alu/trailing/mod.rs

```diff
@@ -0,0 +1,398 @@
+use crate::{
+    air::{SP1CoreAirBuilder, WordAirBuilder},
+    utils::pad_rows_fixed,
+};
+use core::{
+    borrow::{Borrow, BorrowMut},
+    mem::size_of,
+};
+use hashbrown::HashMap;
+use p3_air::{Air, AirBuilder, BaseAir};
+use p3_field::{AbstractField, PrimeField32};
+use p3_matrix::{dense::RowMajorMatrix, Matrix};
+use rwasm::Opcode::{I32Clz, I32Ctz};
+use rwasm_executor::{
+    events::{ByteLookupEvent, ByteRecord},
+    ByteOpcode, ExecutionRecord, Program, DEFAULT_PC_INC,
+};
+use sp1_derive::AlignedBorrow;
+use sp1_stark::{air::MachineAir, Word};
+
+pub const NUM_TRAILING_COLS: usize = size_of::<TrailingCols<u8>>();
+
+#[derive(AlignedBorrow, Default, Debug, Clone, Copy)]
+#[repr(C)]
+pub struct TrailingCols<T> {
+    pub pc: T,
+    pub a: Word<T>,
+    pub b_bytes: [T; 4],
+    // Unified per-op halves:
+    //  - CTZ rows:  [ctz_low16,  ctz_high16]
+    //  - CLZ rows:  [clz_high16, clz_low16]
+    pub half_word_z: [T; 2],
+    pub is_ctz: T,
+    pub is_clz: T,
+    // Minimal boolean flags (no inverse witnesses)
+    pub ctz_low_is_16: T,  // used only on CTZ rows
+    pub clz_high_is_16: T, // used only on CLZ rows
+}
+
+#[derive(Default)]
+pub struct TrailingChip;
+
+impl TrailingChip {
+    fn event_to_row<F: PrimeField32>(
+        &self,
+        event: &rwasm_executor::events::AluEvent,
+        cols: &mut TrailingCols<F>,
+        blu: &mut impl ByteRecord,
+    ) {
+        cols.pc = F::from_canonical_u32(event.pc);
+        cols.a = event.a.into();
+        cols.is_ctz = F::from_bool(event.opcode == I32Ctz);
+        cols.is_clz = F::from_bool(event.opcode == I32Clz);
+
+        let b_val = event.b;
+        let b_bytes = b_val.to_le_bytes();
+        for i in 0..4 {
+            cols.b_bytes[i] = F::from_canonical_u8(b_bytes[i]);
+        }
+
+        // halves
+        let low_half: u16 = (b_val & 0xFFFF) as u16;
```

Comment:
_b_val as u16_ will have the same result
the same for high just use _(b_val >> 16)_